File: src/main/java/com/lessmarkup/userinterface/model/structure/NodeCacheImpl.java
package com.lessmarkup.userinterface.model.structure;

import com.google.gson.JsonElement;
import com.google.inject.Inject;
import com.lessmarkup.Constants;
import com.lessmarkup.TextIds;
import com.lessmarkup.dataobjects.Node;
import com.lessmarkup.dataobjects.NodeAccess;
import com.lessmarkup.framework.helpers.*;
import com.lessmarkup.framework.system.RequestContextHolder;
import com.lessmarkup.interfaces.cache.AbstractCacheHandler;
import com.lessmarkup.interfaces.cache.DataCache;
import com.lessmarkup.interfaces.data.DomainModel;
import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.module.ModuleProvider;
import com.lessmarkup.interfaces.security.CurrentUser;
import com.lessmarkup.interfaces.structure.CachedNodeAccess;
import com.lessmarkup.interfaces.structure.CachedNodeInformation;
import com.lessmarkup.interfaces.structure.ChildHandlerSettings;
import com.lessmarkup.interfaces.structure.GetNodeHandlerPreprocess;
import com.lessmarkup.interfaces.structure.NodeAccessType;
import com.lessmarkup.interfaces.structure.NodeCache;
import com.lessmarkup.interfaces.structure.NodeHandler;
import com.lessmarkup.interfaces.structure.Tuple;
import com.lessmarkup.interfaces.system.SiteConfiguration;
import com.lessmarkup.userinterface.nodehandlers.DefaultRootNodeHandler;
import com.lessmarkup.userinterface.nodehandlers.configuration.ConfigurationRootNodeHandler;
import com.lessmarkup.userinterface.nodehandlers.globalconfiguration.LoginNodeHandler;
import com.lessmarkup.userinterface.nodehandlers.globalconfiguration.DatabaseConfigurationNodeHandler;
import com.lessmarkup.userinterface.nodehandlers.user.ForgotPasswordNodeHandler;
import com.lessmarkup.userinterface.nodehandlers.user.UserCardRecordsNodeHandler;
import com.lessmarkup.userinterface.nodehandlers.user.UserProfileNodeHandler;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Optional;
import java.util.OptionalLong;
import java.util.logging.Level;

public class NodeCacheImpl extends AbstractCacheHandler implements NodeCache {

    private final DomainModelProvider domainModelProvider;
    private final ModuleProvider moduleProvider;
    private final DataCache dataCache;
    private CachedNodeInformation rootNode;
    private final List<CachedNodeInformation> cachedNodes = new ArrayList<>();
    private final HashMap<Long, CachedNodeInformation> idToNode = new HashMap<>();

    @Inject
    public NodeCacheImpl(DomainModelProvider domainModelProvider, ModuleProvider moduleProvider, DataCache dataCache) {
        super(new Class<?>[] {Node.class });
        this.domainModelProvider = domainModelProvider;
        this.moduleProvider = moduleProvider;
        this.dataCache = dataCache;
    }
    
    @Override
    public void initialize(OptionalLong objectId) {
        if (objectId.isPresent()) {
            throw new IllegalArgumentException();
        }
        
        List<CachedNodeInformationImpl> cachedNodesImpl = new ArrayList<>();
        
        try (DomainModel domainModel = this.domainModelProvider.create()) {
            
            domainModel.query()
                    .from(Node.class)
                    .orderBy("position")
                    .toListJava(Node.class)
                    .stream()
                    .map(n -> new CachedNodeInformationImpl(n) {})
                    .forEach(cachedNodesImpl::add);
            
            HashMap<Long, CachedNodeInformation> nodeMap = new HashMap<>();
            
            cachedNodesImpl.forEach(n -> nodeMap.put(n.getNodeId(), n));
            
            domainModel.query()
                    .from(NodeAccess.class)
                    .toListJava(NodeAccess.class)
                    .forEach(na -> {
                        CachedNodeInformation node = nodeMap.get(na.getNodeId());
                        if (node != null) {
                            node.getAccessList().add(new CachedNodeAccess(na));
                        }
                    });
        } catch (Exception ex) {
            LoggingHelper.getLogger(getClass()).log(Level.SEVERE, null, ex);
        }
        
        Optional<CachedNodeInformationImpl> root = cachedNodesImpl.stream().filter(n -> !n.getParentNodeId().isPresent()).findFirst();
        
        CachedNodeInformationImpl rootImpl;
        
        if (root.isPresent()) {
            rootImpl = root.get();
        } else {
            rootImpl = new CachedNodeInformationImpl();
            CachedNodeAccess rootAccess = new CachedNodeAccess();
            rootAccess.setAccessType(NodeAccessType.READ);
            rootImpl.getAccessList().add(rootAccess);
            rootImpl.setNodeId(1);
            rootImpl.setTitle("Home");
            rootImpl.setHandlerId(DefaultRootNodeHandler.class.getSimpleName());
            rootImpl.setHandlerType(DefaultRootNodeHandler.class);
            rootImpl.setVisible(true);
            cachedNodesImpl.add(rootImpl);
        }
        
        this.rootNode = rootImpl;
        
        initializeTree(rootImpl, cachedNodesImpl);
        
        rootImpl.setRoot(this.rootNode);
        
        addVirtualNode(ConfigurationRootNodeHandler.class, 
                Constants.NodePathConfiguration(),
                LanguageHelper.getText(Constants.ModuleTypeMain(), TextIds.CONFIGURATION),
                Constants.ModuleTypeMain(),
                NodeAccessType.NO_ACCESS);
        
        SiteConfiguration siteConfiguration = this.dataCache.get(SiteConfiguration.class);
        
        String adminLoginPage = siteConfiguration.getAdminLoginPage();
        if (adminLoginPage == null || adminLoginPage.length() == 0) {
            adminLoginPage = RequestContextHolder.getContext().getEngineConfiguration().getAdminLoginPage();
        }
        
        if (StringHelper.isNullOrEmpty(adminLoginPage)) {
            adminLoginPage = Constants.NodePathAdminLoginDefaultPage();
        }
        
        addVirtualNode(LoginNodeHandler.class,
                adminLoginPage,
                LanguageHelper.getText(Constants.ModuleTypeMain(), TextIds.ADMINISTRATOR_LOGIN),
                Constants.ModuleTypeMain(),
                NodeAccessType.READ);
        
        CachedNodeInformationImpl node = addVirtualNode(UserProfileNodeHandler.class,
                Constants.NodePathProfile(),
                LanguageHelper.getText(Constants.ModuleTypeMain(), TextIds.USER_PROFILE),
                Constants.ModuleTypeMain(),
                NodeAccessType.READ);
        
        node.setLoggedIn(true);
        
        if (siteConfiguration.getHasUsers()) {
            addVirtualNode(UserCardRecordsNodeHandler.class,
                    Constants.NodePathUserCards(),
                    LanguageHelper.getText(Constants.ModuleTypeMain(), TextIds.USER_CARDS),
                    Constants.ModuleTypeMain(),
                    NodeAccessType.READ);
            addVirtualNode(ForgotPasswordNodeHandler.class,
                    Constants.NodePathForgotPassword(),
                    LanguageHelper.getText(Constants.ModuleTypeMain(), TextIds.FORGOT_PASSWORD),
                    Constants.ModuleTypeMain(),
                    NodeAccessType.READ);
        }
    }
    
    private void initializeTree(CachedNodeInformationImpl node, List<CachedNodeInformationImpl> allNodes) {
        if (node.getHandlerId() != null && node.getHandlerId().length() > 0) {
            Tuple<Class<? extends NodeHandler>,String> handler = this.moduleProvider.getNodeHandler(node.getHandlerId());
            if (handler != null) {
                node.setHandlerType(handler.getValue1());
                node.setHandlerModuleType(handler.getValue2());
                
            }
        }
        
        node.setRoot(this.rootNode);
        
        if (node.getParent() == null) {
            // root
            node.setPath("");
            node.setFullPath("/");
        } else {
            node.setPath(node.getPath().trim().toLowerCase());
            
            if (node.getPath().length() == 0) {
                return;
            }
            
            node.setFullPath(node.getParent() == this.rootNode ? ("/" + node.getPath()) : (node.getParent().getFullPath() + "/" + node.getPath()));
        }
        
        this.cachedNodes.add(node);
        this.idToNode.put(node.getNodeId(), node);
        
        for (CachedNodeInformationImpl child : allNodes) {
            if (!child.getParentNodeId().isPresent() || child.getParentNodeId().getAsLong() != node.getNodeId() || !node.isEnabled()) {
                continue;
            }
            node.getChildren().add(child);
            child.setParent(node);
            initializeTree(child, allNodes);
        }
    }
    
    private <T extends NodeHandler> CachedNodeInformationImpl addVirtualNode(Class<T> type, String path, String title, String moduleType, NodeAccessType accessType) {
        Optional<Long> nodeId = this.idToNode.keySet().stream().max(Long::compare);
        
        CachedNodeInformationImpl node = new CachedNodeInformationImpl();
        node.setNodeId(nodeId.orElse(0l) + 1);
        node.setFullPath("/" + path.toLowerCase());
        node.setPath(path.toLowerCase());
        node.setHandlerModuleType(moduleType);
        node.setParentNodeId(OptionalLong.of(this.rootNode.getNodeId()));
        node.setParent(this.rootNode);
        node.setTitle(title);
        node.setHandlerType(type);
        node.setHandlerId(path);
        node.setRoot(this.rootNode);
        node.setVisible(false);
        
        CachedNodeAccess cachedNodeAccess = new CachedNodeAccess();
        cachedNodeAccess.setAccessType(accessType);
        node.getAccessList().add(cachedNodeAccess);
        
        this.rootNode.getChildren().add(node);
        this.idToNode.put(node.getNodeId(), node);
        
        return node;
    }

    @Override
    public CachedNodeInformation getNode(long nodeId) {
        return idToNode.get(nodeId);
    }

    @Override
    public Tuple<CachedNodeInformation, String> getNode(String path) {
        List<String> nodeParts = new ArrayList<>();
        if (path != null) {
            Arrays.stream(path.split("/")).map(String::trim).filter(p -> p.length() > 0).forEach(nodeParts::add);
        }
        
        if (nodeParts.isEmpty()) {
            return new Tuple<>(this.rootNode, "");
        }
        
        CachedNodeInformation node = this.rootNode;
        
        int pos;
        
        for (pos = 0; pos < nodeParts.size(); pos++) {
            String pathPart = nodeParts.get(pos);
            
            Optional<CachedNodeInformation> child = node.getChildren().stream().filter(n -> n.getPath().equals(pathPart)).findFirst();
            
            if (!child.isPresent()) {
                break;
            }
            
            node = child.get();
        }
        
        String rest = "";
        
        if (pos < nodeParts.size()) {
            rest = StringHelper.join(rest, nodeParts.stream().skip(pos).iterator());
        }
        
        return new Tuple<>(node, rest);
    }

    @Override
    public CachedNodeInformation getRootNode() {
        return this.rootNode;
    }

    @Override
    public List<CachedNodeInformation> getNodes() {
        return this.cachedNodes;
    }
    
    private boolean traverseParents(CachedNodeInformation node, GetNodeHandlerPreprocess filter) {
        if (node.getParent() != null) {
            if (traverseParents(node.getParent(), filter)) {
                return true;
            }
        }
        
        return filter.preprocess(null, node.getTitle(), node.getFullPath(), null, OptionalLong.of(node.getNodeId()));
    }

    @Override
    public NodeHandler getNodeHandler(String path, GetNodeHandlerPreprocess filter) {
        if (path != null) {
            try {
                path = URLDecoder.decode(path, "UTF-8");
            } catch (UnsupportedEncodingException ex) {
                LoggingHelper.getLogger(getClass()).log(Level.SEVERE, null, ex);
            }
        }
        
        if (path != null) {
            int queryPost = path.indexOf('?');
            if (queryPost >= 0) {
                path = path.substring(0, queryPost);
            }
        }
        
        Tuple<CachedNodeInformation, String> foundNode = getNode(path);
       
        if (foundNode == null) {
            LoggingHelper.getLogger(getClass()).log(Level.INFO, "Cannot get node for path ''{0}''", path);
            return null;
        }
        
        CachedNodeInformation node = foundNode.getValue1();

        CurrentUser currentUser = RequestContextHolder.getContext().getCurrentUser();

        if (node.isLoggedIn() && !currentUser.getUserId().isDefined()) {
            LoggingHelper.getLogger(getClass()).info("This node requires user to be logged in");
            return null;
        }

        LoggingHelper.getLogger(getClass()).info("Checking node access rights");
        
        NodeAccessType accessType = node.checkRights(currentUser);
        
        if (accessType == NodeAccessType.NO_ACCESS) {
            LoggingHelper.getLogger(getClass()).info("Current user has no access to specified node");
            return null;
        }
        
        if (filter != null && node.getParent() != null) {
            traverseParents(node.getParent(), filter);
        }
        
        if (node.getHandlerType() == null) {
            LoggingHelper.getLogger(getClass()).log(Level.WARNING, "Node handler is not set for node path ''{0}''", path);
            return null;
        }

        Class<? extends NodeHandler> handlerType = node.getHandlerType();

        if (handlerType.equals(DefaultRootNodeHandler.class) && StringHelper.isNullOrEmpty(RequestContextHolder.getContext().getEngineConfiguration().getConnectionString())) {
            handlerType = DatabaseConfigurationNodeHandler.class;
        }
        
        NodeHandler nodeHandler = DependencyResolver.resolve(handlerType);
        
        if (nodeHandler == null) {
            return null;
        }
        
        String currentTitle = node.getTitle();
        String currentPath = node.getFullPath();
        
        String settingsText = node.getSettings();
        
        JsonElement settingsElement = null;
        
        if (settingsText != null) {
            settingsElement = JsonSerializer.deserializeToTree(settingsText);
        }
        
        nodeHandler.initialize(OptionalLong.of(node.getNodeId()), 
                settingsElement != null && settingsElement.isJsonObject() ? settingsElement.getAsJsonObject() : null, 
                node.getPath(), node.getFullPath(), accessType);
        
        boolean first = true;
        
        String rest = foundNode.getValue2();
        
        while (rest != null && rest.length() > 0) {
            if (filter != null && filter.preprocess(nodeHandler, currentTitle, currentPath, rest, first ? OptionalLong.of(node.getNodeId()) : null)) {
                return null;
            }
            
            first = false;
            
            ChildHandlerSettings childSettings = nodeHandler.getChildHandler(rest);
            
            if (childSettings == null) {
                return null;
            }
            
            nodeHandler = childSettings.getHandler();
            
            currentTitle = childSettings.getTitle();
            currentPath += "/" + childSettings.getPath();
            
            if (childSettings.getRest() == null || childSettings.getRest().length() == 0) {
                break;
            }
            
            rest = childSettings.getRest();
        }
        
        if (filter != null && filter.preprocess(nodeHandler, currentTitle, currentPath, rest, first ? OptionalLong.of(node.getNodeId()) : OptionalLong.empty())) {
            return null;
        }
        
        return nodeHandler;
    }
}

File: src/main/scala/com/lessmarkup/userinterface/model/structure/NodeCacheImpl.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.userinterface.model.structure

import java.io.UnsupportedEncodingException
import java.net.URLDecoder
import java.util.logging.Level

import com.google.inject.Inject
import com.lessmarkup.dataobjects.{Node, NodeAccess}
import com.lessmarkup.framework.helpers._
import com.lessmarkup.framework.nodehandlers.NodeHandlerConfiguration
import com.lessmarkup.framework.system.RequestContextHolder
import com.lessmarkup.interfaces.annotations.NodeAccessType
import com.lessmarkup.interfaces.cache.{AbstractCacheHandler, DataCache}
import com.lessmarkup.interfaces.data.{DomainModel, DomainModelProvider}
import com.lessmarkup.interfaces.module.ModuleProvider
import com.lessmarkup.interfaces.security.CurrentUser
import com.lessmarkup.interfaces.structure._
import com.lessmarkup.interfaces.system.SiteConfiguration
import com.lessmarkup.userinterface.nodehandlers.configuration.ConfigurationRootNodeHandlerFactory
import com.lessmarkup.userinterface.nodehandlers.globalconfiguration.{DatabaseConfigurationNodeHandlerFactory, LoginNodeHandlerFactory}
import com.lessmarkup.userinterface.nodehandlers.user.{ForgotPasswordNodeHandlerFactory, UserCardRecordsNodeHandlerFactory, UserProfileNodeHandlerFactory}
import com.lessmarkup.userinterface.nodehandlers.{DefaultRootNodeHandler, DefaultRootNodeHandlerFactory}
import com.lessmarkup.{Constants, TextIds}

class NodeCacheImpl @Inject() (domainModelProvider: DomainModelProvider, moduleProvider: ModuleProvider, dataCache: DataCache) extends AbstractCacheHandler(Seq(classOf[Node])) with NodeCache {

  private def buildChildNode(dataNode: Node, access: Option[Seq[CachedNodeAccess]],
                             allDataNodes: Seq[Node],
                             allNodeAccess: Map[Long, Seq[CachedNodeAccess]],
                             fullPath: String): CachedNodeInformation = {

    val handler = moduleProvider.getNodeHandler(dataNode.handlerId)
    val childrenFullPath = fullPath + "/" + dataNode.path
    val childrenDataNodes = allDataNodes.filter(n => n.parentId.isDefined && n.parentId.get == dataNode.id)
    val childrenNodes = childrenDataNodes.map(n => buildChildNode(n, allNodeAccess.get(n.id), allDataNodes, allNodeAccess, childrenFullPath))
    new CachedNodeInformationImpl(dataNode, access, childrenFullPath, childrenNodes, false, handler)
  }

  private val cachedNodes: Seq[CachedNodeInformation] = {
    val domainModel: DomainModel = this.domainModelProvider.create

    try {
      val cachedNodeAccess = domainModel.query
        .from(classOf[NodeAccess])
        .toList(classOf[NodeAccess])
        .map(n => new CachedNodeAccess(n))
        .groupBy(na => na.getNodeId)

      val nodes = domainModel.query
        .from(classOf[Node])
        .toList(classOf[Node])

      val existingRoot = nodes.find(n => n.parentId.isEmpty)

      val root: Node = if (existingRoot.isEmpty) {
        val node = new Node
        node.id = 1
        node.title = "Home"
        node.description = ""
        node.settings = None
        node.handlerId = classOf[DefaultRootNodeHandler].getSimpleName
        node.enabled = true
        node.path = ""
        node
      } else {
        existingRoot.get
      }

      val rootChildren = nodes
        .filter(n => (n.parentId.isDefined && n.parentId.get == root.id) || (n.parentId.isEmpty && n.id != root.id))
        .map(c => buildChildNode(c, cachedNodeAccess.get(c.id), nodes, cachedNodeAccess, root.path))

      val configurationNode = createVirtualNode(classOf[ConfigurationRootNodeHandlerFactory], Constants.NodePathConfiguration,
        LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.CONFIGURATION), Constants.ModuleTypeMain, NodeAccessType.NO_ACCESS)

      val siteConfiguration: SiteConfiguration = this.dataCache.get(classOf[SiteConfiguration])

      val adminLoginPage = if (StringHelper.isNullOrWhitespace(siteConfiguration.adminLoginPage)) {
        val engineAdminLoginPage = RequestContextHolder.getContext.getEngineConfiguration.getAdminLoginPage
        if (StringHelper.isNullOrWhitespace(engineAdminLoginPage)) {
          Constants.NodePathAdminLoginDefaultPage
        } else {
          engineAdminLoginPage
        }
      } else {
        siteConfiguration.adminLoginPage
      }

      val adminLoginNode = createVirtualNode(classOf[LoginNodeHandlerFactory], adminLoginPage, LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.ADMINISTRATOR_LOGIN), Constants.ModuleTypeMain, NodeAccessType.READ)

      val userProfileNode = createVirtualNode(classOf[UserProfileNodeHandlerFactory], Constants.NodePathProfile, LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.USER_PROFILE), Constants.ModuleTypeMain, NodeAccessType.READ, loggedIn = true)

      val userNodes: Seq[CachedNodeInformationImpl] = if (!siteConfiguration.hasUsers) Seq() else {
        val userCardsNode = createVirtualNode(classOf[UserCardRecordsNodeHandlerFactory], Constants.NodePathUserCards, LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.USER_CARDS), Constants.ModuleTypeMain, NodeAccessType.READ)
        val forgotPasswordNode = createVirtualNode(classOf[ForgotPasswordNodeHandlerFactory], Constants.NodePathForgotPassword, LanguageHelper.getText(Constants.ModuleTypeMain, TextIds.FORGOT_PASSWORD), Constants.ModuleTypeMain, NodeAccessType.READ)
        Seq(userCardsNode, forgotPasswordNode)
      }

      val rootChildrenWithVirtualNodes = rootChildren ++ Seq(configurationNode, adminLoginNode, userProfileNode) ++ userNodes

      val rootNode = new CachedNodeInformationImpl(root, cachedNodeAccess.get(root.id), root.path, rootChildrenWithVirtualNodes)

      def flatNodesTree(node: CachedNodeInformation): Seq[CachedNodeInformation] =
        node.children.flatMap(c => flatNodesTree(c)) :+ node

      val ret = flatNodesTree(rootNode)

      ret.foreach {
        case n: CachedNodeInformationImpl =>
          n.root = rootNode
          n.children.foreach {
            case c: CachedNodeInformationImpl =>
              c.parent = Option(n)
          }
      }

      ret

    } finally {
      if (domainModel != null) domainModel.close()
    }
  }

  private val rootNode = cachedNodes.find(n => n.parent.isEmpty)

  private val idToNode = cachedNodes.map(n => (n.nodeId, n)).toMap

  private def createVirtualNode[T <: NodeHandlerFactory](handlerType: Class[T], path: String, title: String,
    moduleType: String, accessType: NodeAccessType, loggedIn: Boolean = false): CachedNodeInformationImpl = {

    val nodeId = idToNode.keys.max

    val node = new Node
    node.id = nodeId
    node.path = path.toLowerCase
    node.title = title
    node.handlerId = handlerType.getSimpleName
    node.enabled = true
    node.description = ""
    node.settings = None
    node.addToMenu = false

    val cachedNodeAccess = new CachedNodeAccess(accessType, None, None, nodeId)

    new CachedNodeInformationImpl(node, Option(Seq(cachedNodeAccess)), path, Seq(), loggedIn, Option((handlerType, Constants.ModuleTypeMain)))
  }

  def getNode(nodeId: Long): Option[CachedNodeInformation] = idToNode.get(nodeId)

  def getNode(path: String): Option[(CachedNodeInformation, String)] = {

    val nodeParts: Seq[String] = if (StringHelper.isNullOrEmpty(path)) Seq() else path.split("/").filter(p => p.length > 0)

    if (nodeParts.isEmpty) {
      return Option(this.rootNode.get, "")
    }

    def getChildNode(node: CachedNodeInformation, path: String, parts: Seq[String]): Option[(CachedNodeInformation, Seq[String])] = {
      val child = node.children.find(n => n.path == path)
      if (child.isEmpty) {
        None
      } else {
        if (parts.isEmpty || child.get.children.isEmpty) {
          Option((child.get, parts))
        } else {
          getChildNode(child.get, parts.head, parts.tail)
        }
      }
    }

    val node = getChildNode(rootNode.get, nodeParts.head, nodeParts.tail)

    if (node.isDefined) {
      Option((node.get._1, node.get._2.mkString("/")))
    } else {
      None
    }
  }

  def getRootNode: Option[CachedNodeInformation] = {
    rootNode
  }

  def getNodes: Seq[CachedNodeInformation] = {
    cachedNodes
  }

  private def traverseParents(node: CachedNodeInformation, filter: Option[(NodeHandler, String, String, Seq[String], Option[Long]) => Boolean]): Boolean = {
    if (node.parent.isDefined) {
      if (traverseParents(node.parent.get, filter)) {
        return true
      }
    }
    if (filter.isDefined)
      false
    else
      filter.get(null, node.title, node.fullPath, null, Option(node.nodeId))
  }

  def getNodeHandler(path: String, filter: Option[(NodeHandler, String, String, Seq[String], Option[Long]) => Boolean]): Option[NodeHandler] = {

    val pathDecoded: String = if (StringHelper.isNullOrEmpty(path)) "" else {
      try {
        URLDecoder.decode(path, "UTF-8")
      }
      catch {
        case ex: UnsupportedEncodingException =>
          LoggingHelper.getLogger(getClass).log(Level.SEVERE, null, ex)
          ""
      }
    }

    val pathWithoutQuery = if (StringHelper.isNullOrEmpty(pathDecoded)) pathDecoded else {
      val queryPost: Int = pathDecoded.indexOf('?')
      if (queryPost >= 0) {
        pathDecoded.substring(0, queryPost)
      } else {
        pathDecoded
      }
    }

    val foundNode = getNode(pathWithoutQuery)

    if (foundNode.isEmpty) {
      LoggingHelper.getLogger(getClass).log(Level.INFO, "Cannot get node for path ''{0}''", path)
      return None
    }

    val node = foundNode.get._1

    val currentUser: CurrentUser = RequestContextHolder.getContext.getCurrentUser
    if (node.loggedIn && currentUser.getUserId.isEmpty) {
      LoggingHelper.getLogger(getClass).info("This node requires user to be logged in")
      return None
    }
    LoggingHelper.getLogger(getClass).info("Checking node access rights")
    val accessType: NodeAccessType = node.checkRights(currentUser)
    if (accessType eq NodeAccessType.NO_ACCESS) {
      LoggingHelper.getLogger(getClass).info("Current user has no access to specified node")
      return None
    }
    if (filter != null && node.parent.isDefined) {
      traverseParents(node.parent.get, filter)
    }
    if (node.handlerType == null) {
      LoggingHelper.getLogger(getClass).log(Level.WARNING, "Node handler is not set for node path ''{0}''", path)
      return None
    }

    if (node.handlerType.isEmpty) {
      return None
    }

    val handlerType =
      if ((node.handlerType.isDefined && node.handlerType.get == classOf[DefaultRootNodeHandlerFactory]) && StringHelper.isNullOrEmpty(RequestContextHolder.getContext.getEngineConfiguration.getConnectionString)) {
        classOf[DatabaseConfigurationNodeHandlerFactory]
      } else {
        node.handlerType.get
      }

    val nodeHandlerFactory = DependencyResolver(handlerType)
    if (nodeHandlerFactory == null) {
      return None
    }

    val configuration = new NodeHandlerConfiguration(
      objectId = Option(node.nodeId),
      settings = if (node.settings != null) Option(JsonSerializer.deserializeToTree(node.settings).getAsJsonObject) else None,
      accessType = accessType,
      path = node.path,
      fullPath = node.fullPath
    )

    val nodeHandler = nodeHandlerFactory.createNodeHandler(configuration)

    val currentTitle: String = node.title
    val currentPath: String = node.fullPath

    def handleRest(rest: Seq[String], nodeHandler: NodeHandler, first: Boolean, title: String, path: String): Option[(NodeHandler, Seq[String])] = {
      if (filter.isDefined && filter.get(nodeHandler, currentTitle, currentPath, rest, if (first) Option(node.nodeId) else None)) {
        return None
      }

      val childSettings = nodeHandler.createChildHandler(rest)
      if (childSettings.isEmpty) {
        return None
      }

      val childNodeHandler = childSettings.get.handler
      val childTitle = childSettings.get.title
      val childPath = path + "/" + childSettings.get.path
      if (childSettings.get.rest.isEmpty) {
        return Option((childNodeHandler, rest))
      }

      handleRest(childSettings.get.rest, childNodeHandler, first = false, childTitle, childPath)
    }

    val initialRest = foundNode.get._2.split("/")

    val newNodeHandler = handleRest(initialRest, nodeHandler, first = true, currentTitle, currentPath)

    if (newNodeHandler.isEmpty) {
      return None
    }

    if (filter.isDefined && filter.get(newNodeHandler.get._1, currentTitle, currentPath, newNodeHandler.get._2,
        if (initialRest.nonEmpty) Option(node.nodeId) else None)) {
      return None
    }

    Option(newNodeHandler.get._1)
  }
}